optuna/storages/_journal/storage.py

```diff
@@ -0,0 +1,612 @@
+                    "Another study with name '{}' already exists. "
+                    "Please specify a different name, or reuse the existing one "
+                    "by setting `load_if_exists` (for Python API) or "
+                    "`--skip-if-exists` flag (for CLI).".format(study_name)
+                ),
+            )
+            return
+
+        study_id = self._next_study_id
+        self._next_study_id += 1
+
+        self._studies[study_id] = FrozenStudy(
+            study_name=study_name,
+            direction=StudyDirection.NOT_SET,
+            user_attrs={},
+            system_attrs={},
+            study_id=study_id,
+        )
+        self._study_id_to_trial_ids[study_id] = []
+
+    def _apply_delete_study(self, log: Dict[str, Any]) -> None:
+        study_id = log["study_id"]
+
+        if self._study_exists(study_id, log):
+            fs = self._studies.pop(study_id)
+            assert fs._study_id == study_id
+
+    def _apply_set_study_user_attr(self, log: Dict[str, Any]) -> None:
+        study_id = log["study_id"]
+
+        if self._study_exists(study_id, log):
+            assert len(log["user_attr"]) == 1
+            self._studies[study_id].user_attrs.update(log["user_attr"])
+
+    def _apply_set_study_system_attr(self, log: Dict[str, Any]) -> None:
+        study_id = log["study_id"]
+
+        if self._study_exists(study_id, log):
+            assert len(log["system_attr"]) == 1
+            self._studies[study_id].system_attrs.update(log["system_attr"])
+
+    def _apply_set_study_directions(self, log: Dict[str, Any]) -> None:
+        study_id = log["study_id"]
+
+        if not self._study_exists(study_id, log):
+            return
+
+        directions = [StudyDirection(d) for d in log["directions"]]
+
+        current_directions = self._studies[study_id]._directions
+        if current_directions[0] != StudyDirection.NOT_SET and current_directions != directions:
+            self._raise_if_log_issued_by_pid(
+                log,
+                ValueError(
+                    "Cannot overwrite study direction from {} to {}.".format(
+                        current_directions, directions
+                    )
+                ),
+            )
+            return
+
+        self._studies[study_id]._directions = [StudyDirection(d) for d in directions]
+
+    def _apply_create_trial(self, log: Dict[str, Any]) -> None:
+        study_id = log["study_id"]
+
+        if not self._study_exists(study_id, log):
+            return
+
+        trial_id = len(self._trials)
+        distributions = (
+            {}
+            if "distributions" not in log
+            else {k: json_to_distribution(v) for k, v in log["distributions"].items()}
+        )
+        params = (
+            {}
+            if "params" not in log
+            else {
+                k: distributions[k].to_external_repr(param) for k, param in log["params"].items()
+            }
+        )
```

Comment:
[nits] Can we simplify these lines like below?

```suggestion
        distributions = {}
        if "distributions" in log
            distributions = {k: json_to_distribution(v) for k, v in log["distributions"].items()}
        params = {}
        if "params" in log
            params = {k: distributions[k].to_external_repr(p) for k, p in log["params"].items()}
```